src/txn/txn_timestamp.c

```diff
@@ -0,0 +1,299 @@
+/*-
+ * Copyright (c) 2014-2017 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#ifdef HAVE_TIMESTAMPS
+/*
+ * __wt_txn_parse_timestamp --
+ *	Decodes and sets a timestamp.
+ */
+int
+__wt_txn_parse_timestamp(WT_SESSION_IMPL *session,
+     const char *name, uint8_t *timestamp, WT_CONFIG_ITEM *cval)
+{
+	WT_DECL_RET;
+	WT_ITEM ts;
+	char padbuf[2 * TIMESTAMP_SIZE];
+	wt_timestamp_t tsbuf;
+	const char *hexts;
+	size_t hexlen;
+
+	__wt_timestamp_set(timestamp, zero_timestamp);
+
+	if (cval->len == 0)
+		return (0);
+
+	/* Protect against unexpectedly long hex strings. */
+	if (cval->len > 2 * TIMESTAMP_SIZE)
+		WT_RET_MSG(session, EINVAL,
+		    "Failed to parse %s timestamp '%.*s': too long",
+		    name, (int)cval->len, cval->str);
+
```

Comment:
Off-by-one if `TIMESTAMP_SIZE` is odd, `cval->len == 2 * TIMESTAMP_SIZE`, we add a byte to the beginning of `padbuf` then copy `cval->len`.